aequilibrae/project/network/link_type.py

```diff
@@ -0,0 +1,73 @@
+import string
+from aequilibrae.project.database_connection import database_connection
+
+
+class LinkType:
+    """A link_type object represents a single record in the *link_types* table"""
+    __alowed_characters = string.ascii_letters + '_'
+
+    def __init__(self, link_type_id: str) -> None:
+        if link_type_id is None:
+            raise ValueError('Link type IDs cannot be None')
+
+        if len(link_type_id) != 1 or link_type_id not in string.ascii_letters:
+            raise ValueError('Link Type IDs must be a single ascii character')
+        conn = database_connection()
+        curr = conn.cursor()
+
+        curr.execute('pragma table_info(link_types)')
+        table_struct = curr.fetchall()
+        self.__fields = [x[1] for x in table_struct]
+        self.__original__ = {}
+
+        # data for the link_type
+        curr.execute(f"select * from 'link_types' where link_type_id='{link_type_id}'")
```

Comment:
I'm a big fan of id as the id field, rather than "{table_name}_id"

Reply:
I would also consider not having your __init__ hit the database

I like a pattern of __init__ just storing data into the relevant fields - at the moment - this class doesn't feel very "link_type" specific... most of it is very generic - the structure of the linkType table basically defines the fields in this class. I'd consider moving that logic into a generic databaseTableToClass field. 

Also the fact that you run a query per row in the database seems weird... I'd try to run a single query - select * from link_types - and then create a record for each and store all the records in a dict.

Reply:
The idea would be to create a generic class that would be subclassed for each table (whenever necessary)?

On the creation of items, are you suggesting that I generate all LinkType (and Mode, as they are very similar in structure) objects at once (say whenever the 1st one is requested by the user)?

Reply:
I would not sub-class it,  but yeah

```
class TableLoader():
    def load_table(curr, table_name):
        fields = get_table_struct(curr)
        curr.execute(f"select * from '{table_name}'")
        return [dict(zip(fields,row)) for row in curr.fetchall()]
```

Then you can just use that class in your LinkTypes and Modes classes. Prefer has_a to is_a in almost every circumstance.

Reply:
@jamiecook  Take a look at what I did now. I took the core of your suggestions and got the LinkTypes and LinkType classes a lot cleaner and more robust, I Think.

Reply:
Now there is no much reason why there couldn't be a generic class such as 

```
class GeneralObject:
    def __init__(self, data_set: dict) -> None:
        self.__original__ = {}
        for k, v in data_set.items():
            self.__dict__[k] = v
            self.__original__[k] = v
```

I would subclass that in both LinkType and Mode objects, but not sure if it makes enough sense to do it...